### PR TITLE
Nginx DNS Issue: Kibana and ClamAV

### LINKED ISSUE #2839: Spike - Cloud.gov Kibana Reschedule Route Issue
**Description:**
Cloud.gov does maintenance from Tuesday to Thursday. In this time they can bring apps down, or reschedule them to perform the maintenance. When the Kibana app is rescheduled/brought down and then up, the internal Nginx route from the frontend to Kibana is no longer valid. Nginx does not receive the updated state from cloud.gov giving the Kibana app a new IP address. The workaround for this is to restage the frontend. Nginx gets the updated info and all is well. However, we would like to avoid having to do this manually and see if we can somehow tie the state of Kibana reschedule to the frontend a little better.

One thing to note is that this issue only happens with docker apps in cloud.gov. When a buildpack app is restaged, the routing is not messed up. Another example of this is the clamav nginx router. Below are logs confirming the issue:

```
# Occurred Thursday 2/8/24

# Kibana Logs
14:47:25.073: [CELL.0] Container became healthy
14:47:25.224: [API.2] Process is being rescheduled
14:47:29.414: [CELL.0] Cell c51e9a40-dff6-400a-b1b0-c88318aae8f2 stopping instance 611cb987-a12b-47c3-768e-4b7f
14:47:34.785: [CELL/SSHD.0] Exit status 0
14:47:45.784: [CELL.0] Cell c51e9a40-dff6-400a-b1b0-c88318aae8f2 destroying container for instance 611cb987-a12b-47c3-768e-4b7f
14:47:45.921: [PROXY.0] Exit status 137
14:47:46.624: [CELL.0] Cell c51e9a40-dff6-400a-b1b0-c88318aae8f2 successfully destroyed container for instance 611cb987-a12b-47c3-768e-4b7f

# Frontend Logs
16:17:52.985: [APP/PROC/WEB.0] 2024/02/08 23:17:52 [error] 116#0: *232 connect() failed (113: No route to host) while connecting to upstream, client: 107.2.184.249, server: , request: "GET /kibana/ HTTP/1.1", upstream: "http://10.255.134.124:5601/", host: "tdp-frontend-qasp.app.cloud.gov", referrer: "https://tdp-frontend-qasp.app.cloud.gov/data-files"
16:17:52.987: [RTR.4] tdp-frontend-qasp.app.cloud.gov - [2024-02-08T23:17:52.290706978Z] "GET /kibana/ HTTP/1.1" 404 0 555 "https://tdp-frontend-qasp.app.cloud.gov/data-files" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36" "127.0.0.1:5454" "10.10.1.25:61238" x_forwarded_for:"107.2.184.249, 127.0.0.1" x_forwarded_proto:"https" vcap_request_id:"182a34b5-e22a-4535-7cc2-e5c9549fce50" response_time:0.696023 gorouter_time:0.000101 app_id:"66b86b25-3c40-48a9-916e-c5319e51f308" app_index:"0" instance_id:"d9430b8e-a38a-43ed-6eb4-b38e" x_cf_routererror:"-" x_b3_traceid:"182a34b5e22a45357cc2e5c9549fce50" x_b3_spanid:"7cc2e5c9549fce50" x_b3_parentspanid:"-" b3:"182a34b5e22a45357cc2e5c9549fce50-7cc2e5c9549fce50"
16:17:52.985: [APP/PROC/WEB.0] 2024/02/08 23:17:52 [error] 116#0: *232 open() "/home/vcap/app/public/50x.html" failed (2: No such file or directory), client: 107.2.184.249, server: , request: "GET /kibana/ HTTP/1.1", upstream: "http://10.255.134.124:5601/", host: "tdp-frontend-qasp.app.cloud.gov", referrer: "https://tdp-frontend-qasp.app.cloud.gov/data-files"
```

**Acceptance Criteria:** 
_Create a list of functional outcomes that must be achieved to complete this issue_

- [ ] Determine if there is a way to automatically resolve the routing issues
- [ ] Implement solution
- [ ] Reschedule Kibana and verify Frontend can still route to it without restaging it.

**Tasks:**
_Create a list of granular, specific work items that must be completed to deliver the desired outcomes of this issue_ 

- [ ] Determine if it is possible to tie app state together a bit better.
- [ ] reach out to cloud.gov support to get more context on the issue
- [ ] Implement solution if it exists

**Notes:**
_Add additional useful information, such as related issues and functionality that isn't covered by this specific issue, and other considerations that will be helpful for anyone reading this_ 

- Potential workaround
    - restage nginx proxy application after elastic restage
- Low-fi solution
   - automatically restage app after elastic restage or periodically
   - uptime monitor to notify when application goes down
- _Note 3_

**Supporting Documentation:**
_Please include any relevant log snippets/files/screen shots_

- _Doc 1_
- _Doc 2_

**Open Questions:**
_Please include any questions or decisions that must be made before beginning work or to confidently call this issue complete_ 

- _Open Question 1_
- _Open Question 2_

### PR DESCRIPTION
**Pull request by elipe17**
## Summary of Changes
- Updated nginx config files to give certain hostnames a TTL.
  - Applies to the Kibana and the ClamAV hosts
Pull request closes #2839 

## How to Test
- This is deployed in the `raft` environment
  - Login to the env and verify you can see Kibana as an OFA Sys admin. If you still cant see the Kibana tab, follow the instructions [here](https://github.com/raft-tech/TANF-app/pull/2805).
  - Login to cloud.gov and go to the `raft` apps in the dashboard.
  - Restart the Kibana app
  - Verify during restart you cant see Kibana
  - Verify after the restart that you can see Kibana again (Before if you did this, you would have had to restart the frontend to be able to route to Kibana).

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [ ] Determine if there is a way to automatically resolve the routing issues
- [x] Implement solution
- [ ] Reschedule Kibana and verify Frontend can still route to it without restaging it.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

**Review by raftmsohani — COMMENTED** [review]
I was running to issues when deployed to staging env. Could not figure out what the problem was. Maybe we can pair on this during office hour and you can deploy and we check together

**Comment by andrew-jameson** [comment]
@raftmsohani Do you recall where we have documentation about how Nginx works? I think we did so in the ticket regarding CSRF changes we had to make. Do you feel that would be the correct place to update docs or perhaps we need a separate nginx/kibana markdown?

**Comment by raftmsohani** [comment]
> @raftmsohani Do you recall where we have documentation about how Nginx works? I think we did so in the ticket regarding CSRF changes we had to make. Do you feel that would be the correct place to update docs or perhaps we need a separate nginx/kibana markdown?

The markdown for Nginx is in [here](https://github.com/raft-tech/TANF-app/blob/develop/tdrs-frontend/nginx/README.md), I think we should add the notes to the same location.

**Review by andrew-jameson — APPROVED** [review]
Works for me and code looks solid -- only question might be updating some documentation about changes to the proxy.

**Review by raftmsohani — APPROVED** [review]
LGTM! tested together

**Comment by elipe17** [comment]
@ADPennington, Mo and I tested restarting the prod ClamAV in office hours 4/17/2024. The tests were successful, we could submit files before and after restarting the ClamAV. I've attached some screenshots to sort of show the proof of the claims. If you want to test the ClamAV piece yourself, let me know so I can assist you!

Screenshot 2024-04-17 at 1 49 04 PM

prod-clamav

File: tdrs-frontend/.env [hunk 1]
```diff
@@ -37,7 +37,7 @@ REACT_APP_DEBOUNCE_TIME=30000
 REACT_APP_EVENT_THROTTLE_TIME=60000
 
 # Enable the Kibana tab for dev purposes.
-# REACT_APP_DEV_KIBANA=true
+REACT_APP_DEV_KIBANA=true
```
Comment (ADPennington) [inline tdrs-frontend/.env]: @elipe17 will this be reverted prior to merge?

Comment (elipe17) [reply tdrs-frontend/.env]: @ADPennington, yes. I will revert that now and re-request your approval

**Comment by ADPennington** [comment]
> > @raftmsohani Do you recall where we have documentation about how Nginx works? I think we did so in the ticket regarding CSRF changes we had to make. Do you feel that would be the correct place to update docs or perhaps we need a separate nginx/kibana markdown?
> 
> 
> 
> The markdown for Nginx is in [here](https://github.com/raft-tech/TANF-app/blob/develop/tdrs-frontend/nginx/README.md), I think we should add the notes to the same location.

Is the documentation being updated in this ticket? @raftmsohani

**Comment by elipe17** [comment]
> > > @raftmsohani Do you recall where we have documentation about how Nginx works? I think we did so in the ticket regarding CSRF changes we had to make. Do you feel that would be the correct place to update docs or perhaps we need a separate nginx/kibana markdown?
> > 
> > 
> > The markdown for Nginx is in [here](https://github.com/raft-tech/TANF-app/blob/develop/tdrs-frontend/nginx/README.md), I think we should add the notes to the same location.
> 
> Is the documentation being updated in this ticket? @raftmsohani

@ADPennington I will add that now and ping you when done

**Review by ADPennington — APPROVED** [review]
thank you @elipe17 @raftmsohani 🚀 

Please note @lfrohlich @ttran-hub -- this PR resolves an issue in all envs (including prod) where the frontend is unavailable whenever cloud.gov does maintenance between Tues-Thurs. they do not inform when they do this, so it's important to check periodically throughout the day between Tues-Thurs to see if this has occurred. @ttran-hub the **continuous monitoring log** we manage includes information on how to detect and fix this. Noting for awareness since I will be OOO next week, and the next release that includes this solution will not be in prod until I return. Mo/Eric can support as well, if needed.

**Pull request merged**